### PR TITLE
BN-1153 P2P actor handle socket physical connection and close it if it is no longer required

### PR DESCRIPTION
**Pull request by o-aleksandrov-topl**
## Purpose
Connection shall be handled by Actor P2P. Actor P2P shall close socket connection if it is no longer required

## Approach
We notify remote peer about our usage of connection. If connection no longer used by both parties then client receive command to close connection

## Testing
Unit tests + multinode tests

## Tickets
closes #BN-1153

### CODE REVIEW

File: networking/src/main/scala/co/topl/networking/blockchain/BlockchainNetwork.scala [hunk 1]
```diff
@@ -47,7 +49,13 @@ object BlockchainNetwork {
             .toResource
             .flatMap(connectionLeader =>
               BlockchainSocketHandler
-                .make[F](serverF, clientHandler.usePeer)(peer, connectionLeader, socket.reads, socket.writes)
+                .make[F](serverF, clientHandler.usePeer)(
+                  peer,
+                  connectionLeader,
+                  socket.reads,
+                  socket.writes,
+                  socket.isOpen.ifM(socket.endOfOutput >> socket.endOfInput, Applicative[F].unit)
```
Comment (nandotorterolo) [inline networking/src/main/scala/co/topl/networking/blockchain/BlockchainNetwork.scala:57]: maybe `().pure[F]` avoid applicative import

File: networking/src/main/scala/co/topl/networking/blockchain/BlockchainSocketHandler.scala [hunk 2]
```diff
@@ -133,6 +135,14 @@ object BlockchainSocketHandler {
         22: Byte
       )
 
+    val setRemoteAppLevelF =
```
[the remark this answers is not in the document]
Comment (o-aleksandrov-topl) [reply networking/src/main/scala/co/topl/networking/blockchain/BlockchainSocketHandler.scala:138]: yeah it shall be `notify`, thanks

File: networking/src/main/scala/co/topl/networking/fsnetwork/PeerActor.scala [hunk 3]
```diff
@@ -174,6 +179,13 @@ object PeerActor {
     applicationLevel >> networkLevel >> (newState, newState).pure[F]
   }
 
+  private def closeConnection[F[_]: Async: Logger](state: State[F]): F[(State[F], Response[F])] =
```
Comment (nandotorterolo) [inline networking/src/main/scala/co/topl/networking/fsnetwork/PeerActor.scala:182]: there are a lot of concurrent context in this actor, maybe you dont need it, I am not sure I you need to change them in all the places, ignore my comment if I am wrong
- startApplicationLevel
- stopApplicationLevel
- downloadHeaders
- ...

File: networking/src/main/scala/co/topl/networking/fsnetwork/PeersManager.scala [hunk 4]
```diff
@@ -366,44 +365,85 @@ object PeersManager {
...
+    } yield (newState, newState)
+
   private def updateExternalHotPeersList[F[_]: Async: Logger](state: State[F]): F[Unit] =
     for {
       hotPeersServers <- state.peers.getPeersWithPort(PeerState.Hot).map(_.asRemoteAddress).pure[F]
       _               <- Logger[F].debug(s"Going to update hot peers servers $hotPeersServers")
       _               <- state.hotPeersUpdate(hotPeersServers)
     } yield ()
 
+  private def coldPeer[F[_]: Async: Logger](
```
Comment (nandotorterolo) [inline networking/src/main/scala/co/topl/networking/fsnetwork/PeersManager.scala:386]: maybe you don't need a logger here, or log.

File: networking/src/main/scala/co/topl/networking/fsnetwork/PeerActor.scala (hunk 3 above)
Comment (o-aleksandrov-topl) [reply networking/src/main/scala/co/topl/networking/fsnetwork/PeerActor.scala:182]: required for `>>`

File: networking/src/main/scala/co/topl/networking/fsnetwork/PeersManager.scala (hunk 4 above)
Comment (o-aleksandrov-topl) [reply networking/src/main/scala/co/topl/networking/fsnetwork/PeersManager.scala:386]: required for `moveToColdStatus`

**Review by nandotorterolo — APPROVED** [review]
Maybe you can not create the implementation updating just a few changes.
But doing a review of so many files is not easy, especially when your a not familiar with the code.
Approving to unblock you if you are, but Sean's review will be great if you can wait.

**Pull request merged**